Document multitenancy support in GitOps (#85666)
openshift/openshift-docs: modules/gitops-namespace-scoped-instance.adoc

// Module included in the following assemblies:
//
// * multitenancy/multitenancy-support-in-gitops.adoc

:_mod-docs-content-type: CONCEPT
[id="gitops-namespace-scoped-instance_{context}"]
= Namespace-scoped instance (Application delivery instance)

> After you create an Argo CD custom resource (CR) in one of your namespaces, the {gitops-shortname} Operator brings up an Argo CD in this namespace which you can use for application delivery. In its initial state, this instance has permission to deploy resources only in the same namespace it is installed to. However, there are times when organizations have specific requirements to satisfy that require additional tuning.

Reviewer: ```suggestion
After you create an Argo CD custom resource (CR) in one of your namespaces, the {gitops-shortname} Operator starts an Argo CD in this namespace, which you can use for application delivery. In its initial state, this instance has permission to deploy resources only in the same namespace in which it is installed. However, there are times when organizations have specific requirements to satisfy that require additional tuning.
```
I was confused by the last sentence: does "organization" refer to the user who created the Argo CD CR, or to a tenant/additional party for whom the installer needs to tune the Argo CD instance? Also, does "tuning" just mean configuration, or something more specific? I'd suggest changing the last sentence to something like this: "However, you may need to configure the instance to meet your specific requirements."

With the {gitops-shortname} Operator, you can extend the permissions of the Argo CD instance to enable the Argo CD Application Controller to deploy resources into other namespaces apart from where it is installed. 

[NOTE]
====

> The roles that the {gitops-shortname} Operator in the namespace creates are namespace-scoped and only have access to namespace resources. It cannot perform any actions outside of the namespaces it manages.

Reviewer: ```suggestion
The roles that the {gitops-shortname} Operator in the namespace creates are namespace-scoped and only have access to namespace resources. The Operator cannot perform any actions outside of the namespaces it manages.
```

====

.How does this method work?

The {gitops-shortname} Operator allows OpenShift users to instantiate Argo CD instances in their namespaces, as long as they have permission to create Argo CD resources in the `argoproj.io/v1alpha1` or `argoproj.io/v1beta1` API in their namespace. Currently, a namespace-scoped instance has full administrative permissions for one or many namespaces it manages, similar to allowing all verbs for all resources within that namespace.

> For the Argo CD Application Controller to deploy resources into any other namespace, you need Kubernetes roles and role bindings to label and indicate these namespaces that you want a namespace-scoped instance to manage. The {gitops-shortname} Operator supports the use of `argocd.argoproj.io/managed-by` label to automatically create these roles and role bindings. Use this label and set the value to indicate the managed namespaces. Now, with the `argocd.argoproj.io/managed-by` label, when you deploy the {gitops-shortname} Operator in the namespace-scoped instance mode, it creates a set of roles and role bindings in every namespace that the instance manages. 

Reviewer: ```suggestion
For the Argo CD Application Controller to deploy resources into any other namespace, you need Kubernetes roles and role bindings to label and indicate these namespaces that you want a namespace-scoped instance to manage. The {gitops-shortname} Operator supports the use of the `argocd.argoproj.io/managed-by` label to automatically create these roles and role bindings. Use this label and set the value to indicate the managed namespaces. Then, with the `argocd.argoproj.io/managed-by` label, when you deploy the {gitops-shortname} Operator in the namespace-scoped instance mode, it creates a set of roles and role bindings in every namespace that the instance manages. 
```

[IMPORTANT]
====
The `argocd.argoproj.io/managed-by` label only works for namespaces in the same cluster as the {gitops-shortname} Operator. For remote clusters, you must define the permissions manually.
====

> By default, when labeling the namespace using the `managed-by` label, the {gitops-shortname} Operator gives the Argo CD Application Controller permissions equivalent to the Kubernetes default `admin` cluster role for the labeled namespace. However, you can opt to define an alternate cluster role that is used for the controller and server components by using the `CONTROLLER_CLUSTER_ROLE` and `CONTROLLER_SERVER_ROLE` environment variables respectively in the Operator's `Subscription` resource.  When you provide these variables, the Operator will not create a default role in the namespace but rather only create a role binding in the namespace to the corresponding cluster role. It is up to the administrator to create the cluster role, having full control over the permissions.

Reviewer: ```suggestion
By default, when labeling the namespace using the `managed-by` label, the {gitops-shortname} Operator gives the Argo CD Application Controller permissions equivalent to the Kubernetes default `admin` cluster role for the labeled namespace. However, you can opt to define an alternate cluster role that is used for the controller and server components by using the `CONTROLLER_CLUSTER_ROLE` and `CONTROLLER_SERVER_ROLE` environment variables respectively in the Operator's `Subscription` resource.  When you provide these variables, the Operator will not create a default role in the namespace but rather only create a role binding in the namespace of the corresponding cluster role. It is up to the administrator to create the cluster role, having full control over the permissions.
```

[NOTE]
====

> * When defining your role, Argo CD will attempt to interact with all resources. When using this functionality, you must either provide permission to `view`, `get`, and `watch` all resources in your custom cluster role or configure the Argo CD CR to include or exclude specific resources through the `resourceInclusions` or `resourceExclusions` fields defined in the role.

Reviewer: ```suggestion
* When defining your role, Argo CD will attempt to interact with all resources. Therefore, you must either provide permission to `view`, `get`, and `watch` all resources in your custom cluster role, or configure the Argo CD CR to include or exclude specific resources through the `resourceInclusions` or `resourceExclusions` fields defined in the role.
```

// Module included in the following assemblies:
//
// * multitenancy/multitenancy-support-in-gitops.adoc

:_mod-docs-content-type: CONCEPT
[id="gitops-namespace-scoped-instance_{context}"]
= Namespace-scoped instance (Application delivery instance)

After you create an Argo CD custom resource (CR) in one of your namespaces, the {gitops-shortname} Operator brings up an Argo CD in this namespace which you can use for application delivery. In its initial state, this instance has permission to deploy resources only in the same namespace it is installed to. However, there are times when organizations have specific requirements to satisfy that require additional tuning.

With the {gitops-shortname} Operator, you can extend the permissions of the Argo CD instance to enable the Argo CD Application Controller to deploy resources into other namespaces apart from where it is installed. 

[NOTE]
====
The roles that the {gitops-shortname} Operator in the namespace creates are namespace-scoped and only have access to namespace resources. It cannot perform any actions outside of the namespaces it manages.
====

.How does this method work?

> The {gitops-shortname} Operator allows OpenShift users to instantiate Argo CD instances in their namespaces, as long as they have permission to create Argo CD resources in the `argoproj.io/v1alpha1` API in their namespace. Currently, a namespace-scoped instance has full administrative permissions for one or many namespaces it manages, similar to allowing all verbs for all resources within that namespace.

Reviewer 2: or `argoproj.io/v1beta1` . We support both the versions with `v1beta` being the default.

Author: Thanks for the feedback. I have addressed it in the documentation.

====

By default, when labeling the namespace using the `managed-by` label, the {gitops-shortname} Operator gives the Argo CD Application Controller permissions equivalent to the Kubernetes default `admin` cluster role for the labeled namespace. However, you can opt to define an alternate cluster role that is used for the controller and server components by using the `CONTROLLER_CLUSTER_ROLE` and `CONTROLLER_SERVER_ROLE` environment variables respectively in the Operator's `Subscription` resource.  When you provide these variables, the Operator will not create a default role in the namespace but rather only create a role binding in the namespace to the corresponding cluster role. It is up to the administrator to create the cluster role, having full control over the permissions.

[NOTE]
====
* When defining your role, Argo CD will attempt to interact with all resources. When using this functionality, you must either provide permission to `view`, `get`, and `watch` all resources in your custom cluster role or configure the Argo CD CR to include or exclude specific resources through the `resourceInclusions` or `resourceExclusions` fields defined in the role.

> * It is not possible to manage cluster-scoped resources, such as namespaces, custom resource definitions (CRDs), or cluster roles with a namespace-scoped instance.

Author: @gnunn1 I would not, for now. We do not have clarity yet about the number here. See comments [1](https://docs.google.com/document/d/1A5x3vhoy_fbu2R45viXM0P-IxfaO-Pn34QtNZ8d460Y/edit?disco=AAABC4OnDUw) and [2](https://docs.google.com/document/d/1A5x3vhoy_fbu2R45viXM0P-IxfaO-Pn34QtNZ8d460Y/edit?disco=AAABE2IuKSo).
